crates/walrus-core/src/encoding/basic_encoding.rs

```diff
@@ -86,127 +87,267 @@ impl<'a> fmt::Debug for ReedSolomonEncoder<'a> {
+            .as_ref()
+            .expect("we have just set the encoded symbols")
     }
 
     /// Returns the symbol at the given index.
     ///
+    /// # Errors
+    ///
+    /// Returns an [`EncodeError::IncorrectDataLength`] if the length of the data is not equal to
+    /// the product of the number of source symbols and the symbol size.
+    ///
     /// # Panics
     ///
     /// Panics if the index is out of bounds (i.e, `index >= self.n_shards`).
-    pub fn into_symbol(mut self, index: u16) -> Vec<u8> {
+    pub fn get_symbol(&mut self, data: &[u8], index: u16) -> Result<Vec<u8>, EncodeError> {
+        self.check_data_length(data)?;
         assert!(index < self.n_shards.get());
         let n_source_symbols = self.n_source_symbols.get();
-        if index < n_source_symbols {
-            self.source_symbols[usize::from(index)].to_vec()
+        let symbol_size = usize::from(self.symbol_size.get());
+        Ok(if index < n_source_symbols {
+            data[usize::from(index) * symbol_size..(usize::from(index) + 1) * symbol_size].to_vec()
         } else {
-            self.encode()
+            self.encode(data)?
                 .recovery(usize::from(index - n_source_symbols))
                 .expect(
                     "we have added all source symbols and checked the index, so this cannot fail",
                 )
                 .to_vec()
+        })
+    }
+
+    fn check_data_length(&self, data: &[u8]) -> Result<(), EncodeError> {
+        let expected_length =
+            usize::from(self.n_source_symbols.get()) * usize::from(self.symbol_size.get());
+        crate::ensure!(
+            data.len() == expected_length,
+            EncodeError::IncorrectDataLength(expected_length),
+        );
+        Ok(())
+    }
+
+    pub(crate) fn encode(&mut self, data: &[u8]) -> Result<EncoderResult<'_>, EncodeError> {
+        self.check_data_length(data)?;
+        self.reset();
+
+        for symbol in data.chunks_exact(self.symbol_size.get().into()) {
+            self.encoder
+                .add_original_shard(symbol)
+                .expect("cannot fail as we have checked the length of the data");
         }
+
+        Ok(self
+            .encoder
+            .encode()
+            .expect("we have added all source symbols, so this cannot fail"))
     }
 
-    fn encode(&mut self) -> reed_solomon_simd::EncoderResult<'_> {
+    fn reset(&mut self) {
         self.encoder
-            .encode()
-            .expect("we have added all source symbols, so this cannot fail")
+            .reset(
+                Self::reed_solomon_original_count(self.n_source_symbols),
+                Self::reed_solomon_recovery_count(self.n_source_symbols, self.n_shards),
+                Self::reed_solomon_shard_bytes(self.symbol_size),
+            )
+            .expect("cannot fail as we use the same parameters as when the encoder was created");
+    }
+
+    fn reed_solomon_original_count(n_source_symbols: NonZeroU16) -> usize {
```

Comment:
Nit: since the other function is `reed_solomon_recovery_count`, this may be clearer as follows?
```suggestion
    fn reed_solomon_source_count(n_source_symbols: NonZeroU16) -> usize {
```

Reply:
ah i see it's from the terminology of the RS encoder. Fine as is, I guess.

Reply:
The idea was to have functions that give you exactly the input parameters the RS encoder expects. That's where I also took the name. Unless you feel strongly about this, I would keep it as is.